src/lib/Inference.hs

```diff
@@ -1567,16 +1570,46 @@ binderNestAsPiNest arr = \case
   Empty             -> Empty
   Nest (b:>ty) rest -> Nest (PiBinder b ty arr) $ binderNestAsPiNest arr rest
 
+inferRole :: Type o -> Arrow -> InfererM i o ParamRole
+inferRole ty arr = case arr of
+  ClassArrow -> return DictParam
+  _ -> do
+    zonk ty >>= \case
+      TyKind -> return TypeParam
+      ty' -> isData ty' >>= \case
+        True -> return DataParam
+        -- TODO(dougalm): the `False` branch should throw an error but that's
+        -- currently too conservative. e.g. `data RangeFrom q:Type i:q = ...`
+        -- fails because `q` isn't data. We should be able to fix it once we
+        -- have a `Data a` class (see issue #680).
+        False -> return DataParam
```

Comment:
I don't think we can raise the error because the example that fails, `data RangeFrom q:Type i:q = ...`, is in the prelude! The point is that we only ever actually instantiate it with `q` satisfying `Data q`. But we don't have a way to express that yet.